File: sdks/python/src/tesseron/context.py
"""What a handler is told about the invocation it is running, and what it can send back."""

from __future__ import annotations

import asyncio
import json
from typing import Protocol, TypeVar

from pydantic import BaseModel, ValidationError

from . import elicit_schema
from .errors import ActionError, ProtocolError, TesseronErrorCode
from .json_types import JsonObject, JsonValue, as_object, as_string
from .protocol import AgentIdentity, Capabilities, LogLevel, Methods

__all__ = ["ActionContext", "Cancellation", "DetachedChannel", "GatewayChannel"]

AnswerModel = TypeVar("AnswerModel", bound=BaseModel)

MINIMUM_PERCENT = 0
MAXIMUM_PERCENT = 100


class GatewayChannel(Protocol):
    """The connection a running handler talks back through.

    The session implements this. Keeping it a protocol is what lets the context live in
    this module without reaching into the session's private state, and what lets a detached
    context answer honestly instead of hanging.
    """

    def notify(self, method: str, params: JsonValue) -> None:
        """Writes one fire-and-forget notification, dropping it when the socket is gone."""
        ...

    async def call(self, method: str, params: JsonValue) -> JsonValue:
        """Sends one request and waits for the response the gateway correlates by id.

        Raises:
            ProtocolError: when the gateway refuses, or when the transport closes first.
        """
        ...


class DetachedChannel:
    """The channel a context gets when there is no live connection behind it.

    Notifications go nowhere, which is what a fire-and-forget frame does on a closed socket
    anyway, and every request answers ``-32010`` rather than hanging.
    """

    def notify(self, method: str, params: JsonValue) -> None:
        """Drops the notification."""

    async def call(self, method: str, params: JsonValue) -> JsonValue:
        """Refuses immediately, because nothing is going to answer."""
        raise ProtocolError(
            int(TesseronErrorCode.TRANSPORT_CLOSED),
            "this invocation has no gateway connection",
        )


class Cancellation:
    """A cancellation signal shared between the session and one running handler.

    The gateway cancels with a notification rather than a request, so nothing answers
    ``actions/cancel``; the invocation it names answers ``-32001`` instead. A handler that
    ignores this signal still gets its answer replaced and its task cancelled, so long
    handlers should await :meth:`wait` alongside their own work.
    """

    def __init__(self) -> None:
        self._requested = asyncio.Event()

    @property
    def is_cancelled(self) -> bool:
        """Whether cancellation has already been requested."""
        return self._requested.is_set()

    def cancel(self) -> None:
        """Records the cancellation."""
        self._requested.set()

    async def wait(self) -> None:
        """Resolves as soon as cancellation is requested, immediately if it already was."""
        await self._requested.wait()


class ActionContext:
    """Everything one invocation knows and everything it can send while it runs."""

    def __init__(
        self,
        *,
        action_name: str,
        invocation_id: str,
        cancellation: Cancellation,
        channel: GatewayChannel,
        agent_capabilities: Capabilities,
        agent: AgentIdentity,
        origin: str,
        route: str | None = None,
    ) -> None:
        self.action_name = action_name
        self.invocation_id = invocation_id
        self.cancellation = cancellation
        self.agent_capabilities = agent_capabilities
        self.agent = agent
        self.origin = origin
        self.route = route
        self._channel = channel
        self._highest_percent: int | None = None

    @classmethod
    def detached(cls, action_name: str, invocation_id: str = "detached") -> ActionContext:
        """A context with no connection behind it, for exercising a handler from a test."""
        return cls(
            action_name=action_name,
            invocation_id=invocation_id,
            cancellation=Cancellation(),
            channel=DetachedChannel(),
            agent_capabilities=Capabilities.none(),
            agent=AgentIdentity(id="unknown", name="unknown"),
            origin="unknown",
        )

    @property
    def is_cancelled(self) -> bool:
        """Whether the agent has cancelled this invocation."""
        return self.cancellation.is_cancelled

    async def progress(
        self,
        *,
        percent: int | None = None,
        message: str | None = None,
        data: JsonValue = None,
    ) -> None:
        """Streams one progress update to the agent.

        Percent is an integer clamped into 0 to 100 and never allowed to fall below a value
        already sent for this invocation: an agent rendering a progress bar treats a
        backwards jump as a restart, and the message and data are worth more than the
        regression. Message and data travel unchanged.
        """
        params: JsonObject = {"invocationId": self.invocation_id}
        if message is not None:
            params["message"] = message
        if percent is not None:
            params["percent"] = self._raise_ceiling(percent)
        if data is not None:
            params["data"] = data
        self._channel.notify(Methods.PROGRESS, params)

    async def sample(
        self,
        prompt: str,
        *,
        json_schema: JsonValue = None,
        max_tokens: int | None = None,
    ) -> JsonValue:
        """Asks the agent's model to answer ``prompt``.

        Sampling depth is not a field in any Tesseron frame: the gateway owns
        ``maxSamplingDepth`` and answers ``-32008`` itself, so the host forwards the request
        without counting.

        Raises:
            ActionError: ``-32006 SamplingNotAvailable`` when the agent did not negotiate
                sampling, and whatever the gateway answered otherwise.
        """
        if not self.agent_capabilities.sampling:
            raise ActionError.protocol(
                TesseronErrorCode.SAMPLING_NOT_AVAILABLE,
                "the connected agent did not negotiate sampling",
            )
        params: JsonObject = {"invocationId": self.invocation_id, "prompt": prompt}
        if json_schema is not None:
            params["schema"] = json_schema
        if max_tokens is not None:
            params["maxTokens"] = max_tokens
        answer = await self._call(Methods.SAMPLE, params)
        result = as_object(answer)
        if result is None:
            raise ActionError.protocol(
                TesseronErrorCode.HANDLER_ERROR,
                "the gateway sent an unreadable sampling result",
            )
        return result.get("content")

    async def sample_as(
        self,
        model: type[AnswerModel],
        prompt: str,
        *,
        max_tokens: int | None = None,
    ) -> AnswerModel:
        """:meth:`sample`, with the output schema derived from ``model`` and decoded into it.

        A model asked for structured output answers with the JSON as text, so a string
        result is parsed before it is decoded.

        Raises:
            ActionError: everything :meth:`sample` raises, plus ``-32005 HandlerError`` when
                the answer does not decode.
        """
        content = await self.sample(
            prompt,
            json_schema=model.model_json_schema(mode="validation"),
            max_tokens=max_tokens,
        )
        text = as_string(content)
        if text is not None:
            try:
                content = json.loads(text)
            except json.JSONDecodeError as problem:
                raise ActionError.protocol(
                    TesseronErrorCode.HANDLER_ERROR,
                    f"the sampling result was not valid JSON: {problem}",
                    {"raw": text},
                ) from problem
        try:
            return model.model_validate(content)
        except ValidationError as problem:
            raise ActionError.protocol(
                TesseronErrorCode.HANDLER_ERROR,
                f"the sampling result did not match the expected shape: {problem}",
            ) from problem

    async def confirm(self, question: str) -> bool:
        """Asks the user a yes-or-no question through the agent.

        ``True`` only on an explicit accept. A decline, a cancel, and an agent that never
        negotiated elicitation all answer ``False``, which is the safe reading for the
        destructive-operation gates this exists for.

        Raises:
            ActionError: whatever the gateway answered when the prompt itself failed. The
                user's answer is never an error.
        """
        if not self.agent_capabilities.elicitation:
            return False
        answer = await self._request_elicitation(question, elicit_schema.confirmation_schema())
        return answer.get("action") == "accept"

    async def elicit(self, question: str, *, json_schema: JsonValue = None) -> JsonValue | None:
        """Asks the user for structured content through the agent.

        ``None`` on a decline or a cancel. Unlike :meth:`confirm` a missing capability is an
        error, because structured content has no safe default and the handler has to branch
        on it explicitly.

        Raises:
            ActionError: ``-32007 ElicitationNotAvailable`` when the agent did not negotiate
                elicitation, ``-32602 InvalidParams`` when the schema is not one MCP can
                render, and whatever the gateway answered otherwise.
        """
        if not self.agent_capabilities.elicitation:
            raise ActionError.protocol(
                TesseronErrorCode.ELICITATION_NOT_AVAILABLE,
                "the connected agent did not negotiate elicitation",
            )
        schema = elicit_schema.permissive_schema() if json_schema is None else json_schema
        elicit_schema.validate(schema)
        answer = await self._request_elicitation(question, schema)
        if answer.get("action") != "accept":
            return None
        return answer.get("value")

    async def elicit_as(self, model: type[AnswerModel], question: str) -> AnswerModel | None:
        """:meth:`elicit`, with the form schema derived from ``model`` and decoded into it.

        Raises:
            ActionError: everything :meth:`elicit` raises, plus ``-32005 HandlerError`` when
                the accepted answer does not decode.
        """
        value = await self.elicit(question, json_schema=model.model_json_schema(mode="validation"))
        if value is None:
            return None
        try:
            return model.model_validate(value)
        except ValidationError as problem:
            raise ActionError.protocol(
                TesseronErrorCode.HANDLER_ERROR,
                f"the elicited answer did not match the expected shape: {problem}",
            ) from problem

    async def log(
        self,
        message: str,
        *,
        level: LogLevel = LogLevel.INFO,
        meta: JsonObject | None = None,
    ) -> None:
        """Forwards one log line to the agent. Fire-and-forget."""
        params: JsonObject = {
            "invocationId": self.invocation_id,
            "level": str(level),
            "message": message,
        }
        if meta is not None:
            params["meta"] = meta
        self._channel.notify(Methods.LOG, params)

    async def _request_elicitation(self, question: str, schema: JsonValue) -> JsonObject:
        answer = await self._call(
            Methods.ELICIT,
            {"invocationId": self.invocation_id, "question": question, "schema": schema},
        )
        result = as_object(answer)
        if result is None:
            raise ActionError.protocol(
                TesseronErrorCode.HANDLER_ERROR,
                "the gateway sent an unreadable elicitation result",
            )
        return result

    async def _call(self, method: str, params: JsonValue) -> JsonValue:
        try:
            return await self._channel.call(method, params)
        except ProtocolError as refusal:
            code = refusal.named_code or TesseronErrorCode.INTERNAL_ERROR
            raise ActionError.protocol(code, refusal.message, refusal.data) from refusal

    def _raise_ceiling(self, requested: int) -> int:
        """Returns the percent this update may report, and remembers it."""
        bounded = max(MINIMUM_PERCENT, min(MAXIMUM_PERCENT, requested))
        if self._highest_percent is not None and self._highest_percent > bounded:
            bounded = self._highest_percent
        self._highest_percent = bounded
        return bounded

File: sdks/cpp/src/context.cpp
#include <tesseron/context.hpp>

#include <algorithm>
#include <atomic>
#include <chrono>
#include <utility>

#include <boost/asio/post.hpp>
#include <boost/asio/redirect_error.hpp>
#include <boost/asio/steady_timer.hpp>
#include <boost/asio/this_coro.hpp>
#include <boost/asio/use_awaitable.hpp>

#include "elicit_schema.hpp"
#include "invocation_state.hpp"

namespace tesseron {
namespace {

using boost::asio::awaitable;

/// A timer that is never meant to expire, only to be cancelled. Waiting on one
/// is how a coroutine parks until something on the I/O thread wakes it.
std::shared_ptr<boost::asio::steady_timer> parked_timer(
    const boost::asio::any_io_executor& executor) {
  auto timer = std::make_shared<boost::asio::steady_timer>(executor);
  timer->expires_at(std::chrono::steady_clock::time_point::max());
  return timer;
}

awaitable<void> park_until_cancelled(std::shared_ptr<boost::asio::steady_timer> timer) {
  boost::system::error_code ignored;
  co_await timer->async_wait(boost::asio::redirect_error(boost::asio::use_awaitable, ignored));
}

/// The percent this update may report, remembered so the next one cannot fall
/// below it.
int raise_ceiling(detail::InvocationState& state, int requested) {
  const int bounded = std::clamp(requested, 0, 100);
  const std::lock_guard<std::mutex> guard(state.ceiling_guard);
  const int allowed =
      state.highest_percent.has_value() ? std::max(*state.highest_percent, bounded) : bounded;
  state.highest_percent = allowed;
  return allowed;
}

/// What the agent did with an elicitation prompt, plus whatever it sent back.
struct ElicitationOutcome {
  std::string action;
  Json value;
};

awaitable<Result<ElicitationOutcome>> request_elicitation(detail::InvocationState& state,
                                                          std::string question, Json schema) {
  Json params = Json::object();
  params["invocationId"] = state.invocation_id;
  params["question"] = std::move(question);
  params["schema"] = std::move(schema);

  auto answer = co_await state.channel->call(std::string(methods::kElicit), std::move(params));
  if (!answer.ok()) co_return ActionError::from_protocol_error(std::move(answer).error());

  const Json result = std::move(answer).value();
  const auto action = result.is_object() ? result.find("action") : result.end();
  if (action == result.end() || !action->is_string()) {
    co_return ActionError::protocol(TesseronErrorCode::HandlerError,
                                    "the gateway sent an unreadable elicitation result: " +
                                        result.dump());
  }
  const auto value = result.find("value");
  co_return ElicitationOutcome{action->get<std::string>(),
                               value == result.end() ? Json() : *value};
}

}  // namespace

ProgressUpdate& ProgressUpdate::message(std::string message) {
  message_ = std::move(message);
  return *this;
}

ProgressUpdate& ProgressUpdate::percent(int percent) {
  percent_ = percent;
  return *this;
}

ProgressUpdate& ProgressUpdate::data(Json data) {
  data_ = std::move(data);
  return *this;
}

SampleRequest::SampleRequest(std::string prompt) : prompt_(std::move(prompt)) {}

SampleRequest& SampleRequest::json_schema(Json schema) {
  json_schema_ = std::move(schema);
  return *this;
}

SampleRequest& SampleRequest::max_tokens(std::uint32_t max_tokens) {
  max_tokens_ = max_tokens;
  return *this;
}

ElicitRequest::ElicitRequest(std::string question) : question_(std::move(question)) {}

ElicitRequest& ElicitRequest::json_schema(Json schema) {
  json_schema_ = std::move(schema);
  return *this;
}

LogEntry::LogEntry(LogLevel level, std::string message)
    : level_(level), message_(std::move(message)) {}

LogEntry LogEntry::debug(std::string message) { return {LogLevel::Debug, std::move(message)}; }
LogEntry LogEntry::info(std::string message) { return {LogLevel::Info, std::move(message)}; }
LogEntry LogEntry::warn(std::string message) { return {LogLevel::Warn, std::move(message)}; }
LogEntry LogEntry::error(std::string message) { return {LogLevel::Error, std::move(message)}; }

LogEntry& LogEntry::meta(Json meta) {
  meta_ = std::move(meta);
  return *this;
}

ActionContext::ActionContext(std::shared_ptr<detail::InvocationState> state)
    : state_(std::move(state)) {}

const std::string& ActionContext::action_name() const noexcept { return state_->action_name; }

const std::string& ActionContext::invocation_id() const noexcept { return state_->invocation_id; }

const AgentIdentity& ActionContext::agent() const noexcept { return state_->agent; }

const std::string& ActionContext::origin() const noexcept { return state_->origin; }

const std::optional<std::string>& ActionContext::route() const noexcept { return state_->route; }

Capabilities ActionContext::agent_capabilities() const noexcept {
  return state_->agent_capabilities;
}

std::stop_token ActionContext::stop_token() const noexcept {
  return state_->cancellation.get_token();
}

bool ActionContext::cancelled() const noexcept { return state_->cancellation.stop_requested(); }

awaitable<void> ActionContext::wait_for_cancellation() const {
  if (state_->cancellation.stop_requested()) co_return;

  const auto executor = co_await boost::asio::this_coro::executor;
  const auto waker = parked_timer(executor);
  std::stop_callback wake(state_->cancellation.get_token(), [waker, executor] {
    boost::asio::post(executor, [waker] { waker->cancel(); });
  });
  while (!state_->cancellation.stop_requested()) co_await park_until_cancelled(waker);
}

void ActionContext::progress(const ProgressUpdate& update) const {
  Json params = Json::object();
  params["invocationId"] = state_->invocation_id;
  if (update.message().has_value()) params["message"] = *update.message();
  if (update.percent().has_value()) params["percent"] = raise_ceiling(*state_, *update.percent());
  if (update.data().has_value()) params["data"] = *update.data();
  state_->channel->notify(std::string(methods::kProgress), std::move(params));
}

void ActionContext::log(const LogEntry& entry) const {
  Json params = Json::object();
  params["invocationId"] = state_->invocation_id;
  params["level"] = std::string(name_of(entry.level()));
  params["message"] = entry.message();
  if (entry.meta().has_value()) params["meta"] = *entry.meta();
  state_->channel->notify(std::string(methods::kLog), std::move(params));
}

awaitable<Result<Json>> ActionContext::sample(SampleRequest request) const {
  if (!state_->agent_capabilities.sampling) {
    co_return ActionError::protocol(TesseronErrorCode::SamplingNotAvailable,
                                    "the connected agent did not negotiate sampling");
  }

  Json params = Json::object();
  params["invocationId"] = state_->invocation_id;
  params["prompt"] = request.prompt();
  if (request.json_schema().has_value()) params["schema"] = *request.json_schema();
  if (request.max_tokens().has_value()) params["maxTokens"] = *request.max_tokens();

  auto answer = co_await state_->channel->call(std::string(methods::kSample), std::move(params));
  if (!answer.ok()) co_return ActionError::from_protocol_error(std::move(answer).error());

  const Json result = std::move(answer).value();
  if (!result.is_object()) {
    co_return ActionError::protocol(
        TesseronErrorCode::HandlerError,
        "the gateway sent an unreadable sampling result: " + result.dump());
  }
  const auto content = result.find("content");
  co_return content == result.end() ? Json() : *content;
}

awaitable<Result<bool>> ActionContext::confirm(std::string question) const {
  if (!state_->agent_capabilities.elicitation) co_return false;

  auto answer =
      co_await request_elicitation(*state_, std::move(question), detail::confirmation_schema());
  if (!answer.ok()) co_return std::move(answer).error();
  co_return std::move(answer).value().action == "accept";
}

awaitable<Result<std::optional<Json>>> ActionContext::elicit(ElicitRequest request) const {
  if (!state_->agent_capabilities.elicitation) {
    co_return ActionError::protocol(TesseronErrorCode::ElicitationNotAvailable,
                                    "the connected agent did not negotiate elicitation");
  }

  Json schema =
      request.json_schema().has_value() ? *request.json_schema() : detail::permissive_schema();
  if (auto rejection = detail::validate_elicitation_schema(schema)) {
    co_return std::move(*rejection);
  }

  auto answer = co_await request_elicitation(*state_, request.question(), std::move(schema));
  if (!answer.ok()) co_return std::move(answer).error();

  auto outcome = std::move(answer).value();
  if (outcome.action != "accept") co_return std::optional<Json>();
  co_return std::optional<Json>(std::move(outcome.value));
}

awaitable<void> ActionContext::on_application_thread(std::function<void()> work) const {
  const auto executor = co_await boost::asio::this_coro::executor;
  const auto waker = parked_timer(executor);
  const auto finished = std::make_shared<std::atomic<bool>>(false);

  state_->channel->dispatch_to_application(
      [work = std::move(work), waker, finished, executor]() mutable {
        work();
        finished->store(true, std::memory_order_release);
        boost::asio::post(executor, [waker] { waker->cancel(); });
      });

  // A dispatcher that runs the work inline has already finished by the time
  // control gets here, and its wake-up would land on a timer nothing is waiting
  // on. Every other dispatcher can only finish once this thread parks, because
  // the wake-up it posts is handled here.
  while (!finished->load(std::memory_order_acquire)) co_await park_until_cancelled(waker);
}

}  // namespace tesseron
